drivers/adc/adc_nxp_s32_adc_sar.c

```diff
@@ -0,0 +1,458 @@
+/*
+ * Copyright 2023 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/adc.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/irq.h>
+#include <Adc_Sar_Ip_HwAccess.h>
+#include <Adc_Sar_Ip.h>
+#include <Adc_Sar_Ip_Irq.h>
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#include "adc_context.h"
+
+#define DT_DRV_COMPAT nxp_s32_adc_sar
+LOG_MODULE_REGISTER(adc_nxp_s32_adc_sar, CONFIG_ADC_LOG_LEVEL);
+
+/* ADC internal reference voltage in mV */
+#define ADC_NXP_S32_INTERNAL_VREF 3300
+/* Convert channel of group ADC to channel of physical ADC instance */
+#define ADC_NXP_S32_GROUPCHAN_2_PHYCHAN(group, channel)	\
+						(ADC_SAR_IP_HW_REG_SIZE * group + channel)
+
+struct adc_nxp_s32_config {
+	ADC_Type *base;
+	uint8_t group_channel;
+	uint8_t callback_select;
+	Adc_Sar_Ip_ConfigType *adc_cfg;
+	void (*irq_config_func)(const struct device *dev);
+	const struct pinctrl_dev_config *pin_cfg;
+};
+
+struct adc_nxp_s32_data {
+	const struct device *dev;
+	uint8_t instance;
+	struct adc_context ctx;
+	uint16_t *buffer;
+	uint16_t *buf_end;
+	uint16_t *repeat_buffer;
+	uint32_t mask_channels;
+	uint8_t num_channels;
+};
+
+static uint32_t adc_nxp_s32_get_instance(ADC_Type *base)
```

Comment:
instead of having code stored in code memory to be run at runtime init to determine this, maybe we can just use a DT property and determine the value of the instance byte statically?

Reply:
@manuargue what do you think about his suggesttion?

Reply:
@manuargue what do you think about his suggesttion?

Reply:
is not recommended to have the hw instance as a DT prop, please see https://github.com/zephyrproject-rtos/zephyr/pull/49161#discussion_r950157580 we will update the rest of the existing drivers to use the same approach done on this pr

Reply:
If the concern is the code size, MCUX SDK is doing something similar (e.g. https://github.com/zephyrproject-rtos/hal_nxp/blob/master/mcux/mcux-sdk/drivers/dma/fsl_dma.c#L52-L68). We could also inline this function.

Reply:
> Could we use something like the following to get the instance number statically without DT?

I don't have a strong preference. Current solution makes use of `IP_ADC_BASE_PTRS` defined by the device headers, which will work without needed updates for any NXP S32 device.

Reply:
> Could we use something like the following to get the instance number statically without DT?
> 
> ```
> .inst_num = ((DT_INST_REG_ADDR(n) == IP_ADC_0_BASE) ? 0 : 
> 					(DT_INST_REG_ADDR(n) == IP_ADC_1_BASE) ? 1 : 2)
> ```

That's a good idea, the main goal here is to make it statically initialized. The instance number isn't something I would like to have to put in DT either so I agree with @manuargue about not putting it there. So this solution could be the best of both worlds.

Unfortunately, if there is ever an SOC which has a different number than 3 ADC like in your example, the code would have to become like this I think:

```
#ifdef SOC_1
.inst_num = ((DT_INST_REG_ADDR(n) == IP_ADC_0_BASE) ? 0 : 
					(DT_INST_REG_ADDR(n) == IP_ADC_1_BASE) ? 1 : 2)
#elif SOC_2
.inst_num = ((DT_INST_REG_ADDR(n) == IP_ADC_0_BASE) ? 0 : 
					(DT_INST_REG_ADDR(n) == IP_ADC_1_BASE) ? 1 : 
                                                              (DT_INST_REG_ADDR(n) == IP_ADC_2_BASE) ? 2 : 3)
```

because otherwise there will be syntax errors at compile time.

Something less tedious and more simple to make it a bit more scalable is it could be something like 

```
#define IP_ADC_INSTANCE_CHECK(n) ((DT_INST_REG_ADDR(n) == IP_ADC_##n##_BASE) ? n : 0) |
.inst_num = DT_INST_FOREACH_STATUS_OKAY(IP_ADC_INSTANCE_CHECK) | 0,
```

Reply:
this looks good @decsny, but not sure if the macrobatics done here will work because of the nested `DT_INST_FOREACH_STATUS_OKAY`. We need to give it a try

Reply:
@decsny As your suggestion. 
```
#define IP_ADC_INSTANCE_CHECK(n) ((DT_INST_REG_ADDR(n) == IP_ADC_##n##_BASE) ? n : 0) |
...
.inst_num = DT_INST_FOREACH_STATUS_OKAY(IP_ADC_INSTANCE_CHECK) 0,
```
This marco will not work. it expand to` ((DT_INST_REG_ADDR(0) == IP_ADC_0_BASE) ? n : 0) | ((DT_INST_REG_ADDR(1) == IP_ADC_1_BASE) ? n : 0) | ...`

But base on idea. I implemented as:
```
#define ADC_NXP_S32_INSTANCE_CHECK(indx, n)	\
	((DT_INST_REG_ADDR(n) == IP_ADC_##indx##_BASE) ? indx : 0)
#define ADC_NXP_S32_GET_INSTANCE(n)		\
	LISTIFY(ADC_NXP_S32_INSTANCE_COUNT_DEBRACKET, ADC_NXP_S32_INSTANCE_CHECK, (|), n)
```

Reply:
it would be good to do something like this for the other drivers, too, if they have the same issue

Reply:
Yes, we will align on this way for all drivers wherever is applicable